### PR TITLE
Switch profiles from Lua to library interface

### LINKED ISSUE #1974: Switch profiles from Lua to library interface
Instead of relying on Lua for our profile needs I would like to change the interface of `libosrm_extract` to:

```
Extractor::Run(const Profile& profile)
```

```
class Profile
{
    virtual ~Profile() = default;

    virtual ExtractionWay processWay(const osmium::Way&) = 0;
    virtual ExtractionNode processNode(const osmium::Node&) = 0;
    virtual double processSegment(const Segment&) = 0;
}
```

This abstract over calling lua directly. `osrm-extract` can implement a `Profile` that uses all the current scripting code contained in `libosrm_extract` to be backward compatible in the meantime.

Long term goal is to _drop_ Lua as a dependency. Profiles will be moved to the `node-osrm` side and will be implemented in JS.

### PR DESCRIPTION
**Pull request by kkaefer**
There's now an abstracted interface and no direct calls to Lua anymore.

fixes #1974 

/cc @TheMarex @daniel-j-h

### CODE REVIEW

File: include/extractor/scripting_environment.hpp [hunk 1]
```diff
@@ -1,23 +1,50 @@
...
+    virtual void processSegment(const osrm::util::Coordinate &source,
+                                const osrm::util::Coordinate &target,
+                                double distance,
+                                InternalExtractorEdge::WeightData &weight) = 0;
+
+    ProfileProperties properties;
+};
+
 /**
  * Creates a lua context and binds osmium way, node and relation objects and
```
Comment (TheMarex) [inline include/extractor/scripting_environment.hpp]: Doc is a little bit outdated now.

File: include/extractor/scripting_environment_lua.hpp [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+{
+
+struct LuaScriptingContext final : public ScriptingContext
+{
+    LuaScriptingContext() = default;
+    ~LuaScriptingContext() override = default;
+
+    std::unordered_set<std::string> getNameSuffixList() override;
+    std::vector<std::string> getExceptions() override;
+    void setupSources() override;
```
Comment (TheMarex) [inline include/extractor/scripting_environment_lua.hpp]: OSRM ties to use the clang naming convention with capitalized member functions.

`void SetupSources()`

**Comment by TheMarex** [comment]
Looks good. This shouldn't change any functionality and only add a slight overhead for virtual function calls but is much easier to reason about

Feeling good about including this after 5.3.0 since API doesn't change.

File: include/extractor/scripting_environment_lua.hpp (hunk 2 above)
Comment (kkaefer) [inline include/extractor/scripting_environment_lua.hpp]: Changed.

File: include/extractor/scripting_environment.hpp (hunk 1 above)
Comment (kkaefer) [inline include/extractor/scripting_environment.hpp]: Changed

**Comment by TheMarex** [comment]
This is included in master now. :tada:

**Comment by springmeyer** [comment]
I re-read this PR in the contex of #2926. I cannot see any obvious problems that would cause an increase in memory usage. But I notice that locks are used when the `GetLuaContext` is called. Therefore I wonder if `GetLuaContext` were called a more often then more locking could cause less efficient allocation and therefore memory growth. Just a hunch, but going on this hunch I looked in the PR to see if more locks could be happening...

It seems like the answer is yes, but not many more. Would appreciate others eyes to rule this out. For example [here the context was previously accessed once and then used in `util::luaFunctionExists` and `RestrictionParser` directly](https://github.com/Project-OSRM/osrm-backend/pull/2647/files#diff-edf14d92e413123a2b940ab0b8491ffbL111). After this PR the code is much cleaner but the context is accessed once when [`SetupSources()` is called](https://github.com/Project-OSRM/osrm-backend/pull/2647/files#diff-edf14d92e413123a2b940ab0b8491ffbR107) and a second time when [`RestrictionParser` is created (due to the usage of scripting_environment.GetProfileProperties() inside its constructor)](https://github.com/Project-OSRM/osrm-backend/pull/2647/files#diff-edf14d92e413123a2b940ab0b8491ffbR133). Now, creating the context twice rather than once may have no realworld performance penalty at all. But maybe it could? Or maybe there are other spots in the code that more locks are happening that I don't notice?

**Pull request merged**